Sources/MongoSwift/ChangeStream.swift

```diff
@@ -84,40 +91,27 @@ public class ChangeStream<T: Codable>: Sequence, IteratorProtocol {
     /// `maxAwaitTimeMS` milliseconds as specified in the `ChangeStreamOptions`, or for the server default timeout
     /// if omitted.
     public func next() -> T? {
-        // If an error exists, refuse iterating the change stream to avoid overwriting the original error.
-        guard self.error == nil else {
-            return nil
-        }
-        // Allocate space for a reference to a BSON pointer.
-        let out = UnsafeMutablePointer<BSONPointer?>.allocate(capacity: 1)
-        defer {
-            out.deinitialize(count: 1)
-            out.deallocate()
-        }
-
-        guard mongoc_change_stream_next(self.changeStream, out) else {
+        // We already closed the mongoc change stream, either because we reached the end or encountered an error.
+        guard case let .open(_, connection, _, session) = self.state else {
+            self.error = ClosedChangeStreamError
             return nil
         }
-
-        guard let pointee = out.pointee else {
-            fatalError("The change stream was advanced, but the document is nil.")
-        }
-
-        // we have to copy because libmongoc owns the pointer.
-        let doc = Document(copying: pointee)
-
-        // Update the resumeToken with the `_id` field from the document.
-        guard let resumeToken = doc["_id"] as? Document else {
-            self.swiftError =
-                    RuntimeError.internalError(message: "_id field is missing from the change stream document.")
-            return nil
-        }
-        self.resumeToken = ResumeToken(resumeToken)
-
         do {
-            return try decoder.decode(T.self, from: doc)
+            // If an error exists, refuse iterating the change stream to avoid overwriting the original error.
+            guard self.error == nil else {
+                return nil
+            }
+            let operation = ChangeStreamNextOperation(changeStream: self)
+            guard let out = try operation.execute(using: connection, session: session) else {
+                if self.getChangeStreamError() != nil {
```

Comment:
we probably want to be assigning this error (if non nil) to `self.error` before closing.